ethcore/src/private_transactions/mod.rs

```diff
@@ -0,0 +1,907 @@
+					encrypted: encrypted_transaction,
+					contract: contract,
+				};
+				let private_state = self.execute_private_transaction(BlockId::Latest, &signed_transaction)?;
+				trace!("Private transaction created, encrypted transaction: {:?}, private state: {:?}", private, private_state);
+				let contract_validators = self.get_validators(BlockId::Latest, &contract)?;
+				trace!("Required validators: {:?}", contract_validators);
+				let private_state_hash = keccak(&private_state);
+				trace!("Hashed effective private state for sender: {:?}", private_state_hash);
+				self.transactions_for_signing.lock().add_transaction(private.hash(), signed_transaction, contract_validators, private_state)?;
+				self.broadcast_private_transaction(private.rlp_bytes().into_vec());
+				Ok(Receipt {
+					hash: tx_hash,
+					contract_address: None,
+					status_code: 0,
+				})
+			}
+		}
+	}
+
+	/// Try to unlock account using stored passwords
+	fn unlock_account(&self, account: &Address) -> Result<bool, EthcoreError> {
+		let accounts = self.accounts.read().clone().ok_or_else(|| PrivateTransactionError::AccountProviderIsMalformed)?;
+		let accounts = accounts.upgrade().ok_or_else(|| PrivateTransactionError::AccountProviderIsMalformed)?;
+		let passwords = self.config.read().passwords.clone();
+		for password in passwords {
+			if let Ok(()) = accounts.unlock_account_temporarily(account.clone(), password) {
+				return Ok(true);
+			}
+		}
+		Ok(false)
+	}
+
+	/// Extract signed transaction from private transaction
+	fn extract_original_transaction(&self, private: PrivateTransaction, contract: &Address) -> Result<UnverifiedTransaction, EthcoreError> {
+		let encrypted_transaction = private.encrypted.clone();
+		let transaction_bytes = self.decrypt(contract, &encrypted_transaction)?;
+		let original_transaction: UnverifiedTransaction = UntrustedRlp::new(&transaction_bytes).as_val()?;
+		Ok(original_transaction)
+	}
+
+	/// Process received private transaction
+	pub fn import_private_transaction(&self, rlp: &[u8]) -> Result<(), EthcoreError> {
+		let validator_accounts = self.config.read().validator_accounts.clone();
+		trace!("Private transaction received");
+		if validator_accounts.is_empty() {
+			self.broadcast_private_transaction(rlp.into());
+			return Ok(());
+		}
+		let private_tx: PrivateTransaction = UntrustedRlp::new(rlp).as_val()?;
+		let contract = private_tx.contract;
+		let contract_validators = self.get_validators(BlockId::Latest, &contract)?;
+
+		match contract_validators
+			.iter()
+			.filter(|&&address| validator_accounts
+				.iter()
+				.any(|&validator| validator == address))
+			.next() {
+			None => {
+				// Not for verification, broadcast further to peers
+				self.broadcast_private_transaction(rlp.into());
+				return Ok(());
+			},
+			Some(&validation_account) => {
+				trace!("Private transaction taken for verification");
+				let client = self.client.read().clone().ok_or_else(|| PrivateTransactionError::ClientIsMalformed)?;
```

Comment:
Perhpas pass `client` as a parameter instead to avoid circular dependencies? (Similar to how it's done in miner (not saying it's perfect, but I think it's more clear than having `Mutex<Option<Weak<Client>>>`)